File: 200_number_of_islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int H=grid.length;
        if(H==0)return 0;
        
        int W=grid[0].length;
        
        if(W==0) return 0;
        
        int count=0;
        
        
        for(int i=0;i<H;i++){
            for(int j=0;j<W;j++){
                if(grid[i][j]=='1'){
                    count++;
                    dfs(grid,i,j);
                }
            }
        }
        return count;
    }
    
     void dfs(char[][] grid, int i, int j){
          int H=grid.length;
     
        
          int W=grid[0].length;
          grid[i][j]='0';
         
         if(i+1 < H && grid[i+1][j]=='1'){
             dfs(grid,i+1,j);
         }
                 
         if(j+1 < W && grid[i][j+1] =='1'){
             dfs(grid,i,j+1);
         }
                 
                 
        if(i-1 >=0 && grid[i-1][j] =='1'){
             dfs(grid,i-1,j);
        }
            
                  
        if(j-1 >=0 && grid[i][j-1] =='1'){
             dfs(grid,i,j-1);
                 
         }
    
     }  
    
}

File: 200_number_of_islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        H = len(grid)
        W=len(grid[0])
        
        if(H==0 or W==0):
            return 0
        count =0
        
        
        for y in range(0,H):
            for x in range(0,W):
                if(grid[y][x]=='1'):
                    count+=1
                    self.dfs(grid, y,x)
                
        return count
        
    def dfs(self, grid: List[List[str]], y,x):
        H = len(grid)
        W=len(grid[0])
        
        grid[y][x]=0
        
        if(y+1<H and grid[y+1][x]=='1'):
            self.dfs(grid,y+1,x)
            
        if(y-1>=0 and grid[y-1][x]=='1'):
            self.dfs(grid,y-1,x)
            
        if(x+1<W and grid[y][x+1]=='1'):
            self.dfs(grid,y,x+1)
            
        if(x-1>=0 and grid[y][x-1]=='1'):
            self.dfs(grid,y,x-1)
